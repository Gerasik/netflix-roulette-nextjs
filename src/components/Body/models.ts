import * as Models from 'models';
import * as BodyModels from 'containers/Body/models';
import * as SearchModels from 'containers/Search/models';

interface MoviesData {
  data: Models.MoviesDataMap;
  total: number;
  sortBy: string;
}
export interface BodyProps {
  moviesData: MoviesData;
  setStartData: BodyModels.ActionCreator.SetStartData;
  addData: BodyModels.ActionCreator.AddData;
  changeSortBy: SearchModels.ActionCreator.ChangeSortBy;
}
