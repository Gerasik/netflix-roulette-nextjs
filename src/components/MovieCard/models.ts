import * as Models from 'models';

export interface MovieCardProps {
  data: Models.MovieMap;
}
