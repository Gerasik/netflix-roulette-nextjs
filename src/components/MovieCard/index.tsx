import React, { FunctionComponent, useState } from 'react';
import Link from 'next/link';
import _ from 'lodash';

import imagePlaceholder from 'assets/image_placeholder.png';
import styles from './style.module.scss';
import { MovieCardProps } from './models';

const MovieCard: FunctionComponent<MovieCardProps> = ({ data }) => {
  const title = data.get('title');
  const posterPath = data.get('poster_path');
  const releaseDate = data.get('release_date');
  const genresUniq = _.uniq(data.get('genres').toArray());
  const id = data.get('id');

  const [imgError, setImgError] = useState(true);

  const image = imgError ? (
    <img
      src={posterPath}
      alt={title}
      className={styles.image}
      onError={(): void => setImgError(false)}
    />
  ) : (
    <div className={styles.error}>
      <img className={styles['error-image']} src={imagePlaceholder} alt="Error load" />
    </div>
  );

  return (
    <div className={styles.item}>
      {image}
      <div className={styles.container}>
        <Link href="/film/[id]" as={`/film/${id}`}>
          <a className={styles.title}>{title}</a>
        </Link>
        <span className={styles.date}>{new Date(releaseDate).getFullYear()}</span>
      </div>
      <div className={styles['genre-list']}>
        {genresUniq.map(item => {
          return (
            <span key={item} className={styles.genre}>
              {item}
            </span>
          );
        })}
      </div>
    </div>
  );
};

export default MovieCard;
