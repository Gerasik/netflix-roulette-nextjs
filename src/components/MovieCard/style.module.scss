@import '../../styles/colors';

$opacityColor: rgba(
  $color: $fontColor,
  $alpha: 0.7,
);

$opacityColorSecond: rgba(
  $color: $fontColor,
  $alpha: 0.4,
);

.item {
  color: $opacityColor;
  .image,
  .error {
    margin-top: 2em;
    min-height: 36em;
    background-color: rgba($color: $fontColor, $alpha: 0.7);
  }
  .image {
    width: 100%;
  }
  .error {
    display: flex;
    justify-content: center;
    align-items: center;
    .error-image {
      width: 50%;
    }
  }
  .title {
    font-weight: bold;
    color: $fontColor;
  }
  .date {
    padding: 0.5em 1.1em;
    border: 0.05em solid $opacityColorSecond;
    border-radius: 0.8em;
    margin: 0.1em;
  }
  .container {
    display: flex;
    justify-content: space-between;
    margin-top: 0.8em;
    align-items: center;
    min-height: 2.5em;
  }
  .genre-list {
    margin-top: 0.5em;
  }
  .genre {
    display: inline-block;
    color: $opacityColorSecond;
    margin-right: 0.4em;
  }
}
