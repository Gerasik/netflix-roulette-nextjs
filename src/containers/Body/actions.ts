import ActionType from './constants';
import { ActionCreator } from './models';

export const setData: ActionCreator.SetData = payload => ({
  type: ActionType.SET_DATA,
  payload,
});

export const setStartData: ActionCreator.SetStartData = () => ({
  type: ActionType.SET_START_DATA,
});

export const addData: ActionCreator.AddData = () => ({
  type: ActionType.ADD_DATA,
});
