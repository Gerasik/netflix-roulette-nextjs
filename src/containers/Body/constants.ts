const ActionType = {
  SET_DATA: 'SET-DATA',
  SET_START_DATA: 'SET-START-DATA',
  ADD_DATA: 'ADD-DATA',
};

export default ActionType;
