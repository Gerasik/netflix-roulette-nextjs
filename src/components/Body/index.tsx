import React, { useEffect, FunctionComponent } from 'react';
import classNames from 'classnames';
import _ from 'lodash';

import { Action } from 'containers/Search/models';
import MovieCard from 'components/MovieCard';
import styles from './style.module.scss';
import { BodyProps } from './models';

const Body: FunctionComponent<BodyProps> = ({
  setStartData,
  moviesData,
  changeSortBy,
  addData,
}) => {
  const { data, total, sortBy } = moviesData;

  const bodyAction = _.throttle(() => checkAddContent(), 300);

  useEffect(() => {
    document.addEventListener('scroll', bodyAction);
    return (): void => {
      document.removeEventListener('scroll', bodyAction);
    };
  }, [setStartData]);

  function checkAddContent(): void {
    const lengthToBottom = document.body.clientHeight - window.innerHeight - window.scrollY;
    if (lengthToBottom < 100) {
      addData();
    }
  }

  return (
    <div className={styles['movie-container']}>
      <header className={styles['movie-header']}>
        <span className={styles['movie-count']}>{total ? `${total} movie found` : ''}</span>
        <div className={styles['movie-buttons']}>
          <span>sort by</span>
          <button
            type="button"
            className={classNames(styles['sort-release'], {
              [styles.active]: sortBy === 'release_date',
            })}
            onClick={(): Action.ChangeSortBy => changeSortBy('release_date')}
          >
            release date
          </button>
          <button
            type="button"
            className={classNames(styles['sort-rating'], {
              [styles.active]: sortBy === 'vote_average',
            })}
            onClick={(): Action.ChangeSortBy => changeSortBy('vote_average')}
          >
            rating
          </button>
        </div>
      </header>
      {!total ? (
        <div className={styles['movie-empty-result']}>
          <span>No films found</span>
        </div>
      ) : (
        <ul className={styles['movie-list']}>
          {data.map(item => (
            <li key={item.get('id')} className={styles['movie-item']}>
              <MovieCard data={item} />
            </li>
          ))}
          {data && data.size % 3 === 2 ? <li className={styles['movie-item']} /> : null}
        </ul>
      )}
    </div>
  );
};

export default Body;
