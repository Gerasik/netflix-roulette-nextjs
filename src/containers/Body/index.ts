import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';

import Body from 'components/Body';
import { changeSortBy } from 'containers/Search/actions';
import { setStartData, addData } from './actions';
import { moviesData } from './selectors';

const mapStateToProps = createStructuredSelector({
  moviesData,
});

const mapDispatchToProps = {
  changeSortBy,
  setStartData,
  addData,
};

export default connect(mapStateToProps, mapDispatchToProps)(Body);
