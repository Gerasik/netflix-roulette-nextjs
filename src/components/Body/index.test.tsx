/* eslint-disable @typescript-eslint/camelcase */
import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { testMovie1, testMovie2, testMovieResponse } from 'mock';
import MovieCard from 'components/MovieCard';
import Body from './index';

const testList = {
  data: Immutable.fromJS([testMovie1, testMovie2]),
  total: 2,
  sortBy: 'release_date',
};

describe('<Body />', () => {
  const wrapper = shallow(
    <Body
      moviesData={testList}
      setStartData={() => ({ type: 'string', payload: testMovieResponse })}
      changeSortBy={() => ({ type: 'string', payload: 'testMovieResponse' })}
    />
  );

  wrapper.find('.sort-rating').simulate('click');
  wrapper.find('.sort-release').simulate('click');

  it('should render with test data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should be two MovieCard component', () => {
    expect(wrapper.find(MovieCard).length).toBe(2);
  });

  it('right props first MovieCard component', () => {
    expect(wrapper.find(MovieCard).get(0).props.data).toBe(testList.data.get(0));
  });
});

describe('<Body /> test empty data', () => {
  const wrapper = shallow(
    <Body
      moviesData={{ ...testList, total: 0 }}
      setStartData={() => ({ type: 'string', payload: testMovieResponse })}
      changeSortBy={() => ({ type: 'string', payload: 'testMovieResponse' })}
    />
  );

  it('should render with test data', () => {
    expect(wrapper.find('.movie-empty-result').length).toBe(1);
  });
});
