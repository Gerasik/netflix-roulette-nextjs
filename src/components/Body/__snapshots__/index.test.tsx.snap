// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Body /> should render with test data 1`] = `
<div
  className="movie-container"
>
  <header
    className="movie-header"
  >
    <span
      className="movie-count"
    >
      2 movie found
    </span>
    <div
      className="movie-buttons"
    >
      <span>
        sort by
      </span>
      <button
        className="sort-release active"
        onClick={[Function]}
        type="button"
      >
        release date
      </button>
      <button
        className="sort-rating"
        onClick={[Function]}
        type="button"
      >
        rating
      </button>
    </div>
  </header>
  <ul
    className="movie-list"
  >
    <li
      className="movie-item"
      key="1"
    >
      <MovieCard
        data={
          Immutable.Map {
            "budget": 54545,
            "release_date": "2020-05-01",
            "vote_average": 55,
            "tagline": "test",
            "genres": Immutable.List [
              "drama",
              "sitcom",
              "drama",
            ],
            "poster_path": "/movie_url/image.jpg",
            "title": "first movie",
            "runtime": 5454,
            "overview": "ove",
            "vote_count": 66,
            "id": 1,
            "revenue": 5454,
          }
        }
      />
    </li>
    <li
      className="movie-item"
      key="2"
    >
      <MovieCard
        data={
          Immutable.Map {
            "budget": 54545,
            "release_date": "2020-05-01",
            "vote_average": 55,
            "tagline": "test",
            "genres": Immutable.List [
              "drama",
            ],
            "poster_path": "/movie_url/image2.jpg",
            "title": "second movie",
            "runtime": 5454,
            "overview": "ove",
            "vote_count": 66,
            "id": 2,
            "revenue": 5454,
          }
        }
      />
    </li>
  </ul>
</div>
`;
